docs/proposals/01-firmware-update.md

```diff
@@ -0,0 +1,381 @@
+- `update job`: the execution item, which runs concrete implementation of the BIOS/firmware update routine on target hardware server;
+- `scan job`: the execution item, which runs concrete implementation for scanning of the firmware installed on target hardware server;
+- `discovery job`: discovering of the BIOS/firmware versions available for download or installation. Vendor-specific as an `update job`;
+- `update strategy`: the path chosen to apply updates, e.g.: pre-built boot image with updates, docker image with baked updates, vendor-specific CLI tool, etc.
+
+## Motivation
+
+It is necessary to provide a robust, reliable and scalable solution to automate servers' firmware updating process.
+Aside from that, it SHOULD also be as much kubernetes-native as possible.
+It SHOULD provide a clear and concise API.
+It SHOULD provide the ability to automate the update process along with the ability to override common settings in particular circumstances for particular servers.
+
+### Goals
+
+The following list gives general design goals for BIOS/Firmware updates:
+
+- the solution SHOULD be vendor-agnostic aside from concrete update job implementation;
+- the solution SHOULD allow automated hardware servers' firmware lifecycle maintaining;
+- the solution MUST be extensible by the possibility of using plugins for update strategy;
+- the solution MUST be extensible by the possibility of adding vendor-specific update jobs;
+- the solution SHOULD be as kubernetes-native as possible;
+
+### Non-Goals
+
+## Proposal
+
+### Custom resources
+
+The following CRs aimed to represent the current state of a particular server, a desired state of a group of servers and available firmware versions for a particular manufacturer-model:
+
+- [ServerFirmware](#serverfirmware)
+- [ServerFirmwareGroup](#serverfirmwaregroup)
+- [DiscoveredFirmware](#discoveredfirmware)
+
+All the following CRs MUST be cluster-scoped.
+The firmware versions defined for concrete `ServerFirmware` object MUST take precedence over those provided by corresponding `ServerFirmwareGroup` object.
+
+#### ServerFirmware
+
+`ServerFirmware` CR represents the desired state of concrete hardware server.
+The `.spec` of this type references the `Server` object, reflects its `.status.bios` field and contains the list of firmwares desired to be installed.
+The `.status` of this type contains information about the BIOS/firmware versions which are actually installed on the server.
+Aside from that `.spec` contains the scan threshold and the `.status` contains last scan operation timestamp.
+These two fields allow calculating whether the scanning for installed firmware is required or not.
+The `ServerFirmware` object SHOULD be created along with corresponding `Server` object and MUST be unique across the cluster.
+
+```yaml
+apiVersion: metal.ironcore.dev/v1alpha1
+kind: ServerFirmware
+metadata:
+  name: foo
+spec:
+  scanThreshold: 30m
+  serverRef:
+    name: foo
+  bios:
```

Comment:
Since update service will anyways request for server object - at least to get related bmc for access type and credentials, I think it's not necessary to store manufacturer and model in this resource. But I have no strong opinion on that, since it would not affect anything.